codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java

```diff
@@ -0,0 +1,136 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.compression;
+
+import com.aayushatharva.brotli4j.decoder.DecoderJNI;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public final class BrotliDecoder extends ByteToMessageDecoder {
+
+    private enum State {
+        DONE, NEEDS_MORE_INPUT, ERROR
+    }
+
+    private DecoderJNI.Wrapper decoder;
+    private final AtomicBoolean destroyed = new AtomicBoolean();
+
+    static {
+        try {
+            Brotli.ensureAvailability();
+        } catch (Throwable throwable) {
+            throw new ExceptionInInitializerError(throwable);
+        }
+    }
+
+    private State decompress(ByteBuf input, List<Object> output) {
+        for (;;) {
+            switch (decoder.getStatus()) {
+                case DONE:
+                    return State.DONE;
+
+                case OK:
+                    decoder.push(0);
+                    break;
+
+                case NEEDS_MORE_INPUT:
+                    if (decoder.hasOutput()) {
+                        output.add(Unpooled.wrappedBuffer(decoder.pull()));
+                    }
+
+                    if (!input.isReadable()) {
+                        return State.NEEDS_MORE_INPUT;
+                    }
+
+                    ByteBuffer decoderInputBuffer = decoder.getInputBuffer();
+                    decoderInputBuffer.clear();
+                    int readBytes = readBytes(input, decoderInputBuffer);
+                    decoder.push(readBytes);
+                    break;
+
+                case NEEDS_MORE_OUTPUT:
+                    output.add(Unpooled.wrappedBuffer(decoder.pull()));
+                    break;
+
+                default:
+                    return State.ERROR;
+            }
+        }
+    }
+
+    private static int readBytes(ByteBuf in, ByteBuffer dest) {
+        int limit = Math.min(in.readableBytes(), dest.remaining());
+        ByteBuffer slice = dest.slice();
+        slice.limit(limit);
+        in.readBytes(slice);
+        dest.position(dest.position() + limit);
+        return limit;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        decoder = new DecoderJNI.Wrapper(8 * 1024);
```

Comment:
IIUC, the decoder is initialized here to allow reusing it. However, the internal state is tiny, it should be ok to create a new instace instead of removing and adding the existing one. Consider initializing decoder in the constructor, then the `decoder` can be `final`.

Reply:
just a note if we do this in the constructor we will also need to destroy it in a finalizer as you never know if the user ends up adding the handler to the pipeline

Reply:
@normanmaurer That was my original reasoning for making this lazy. However, is this really a concern as this Decoder  is only created when the response has a br content-encoding? see https://github.com/netty/netty/pull/10960/files#diff-2e33eaddd20ffe54e6c4e266fc9cf90ac5f8715ec82a7bcef5d45dbcfdee4996R72

Reply:
Not a concert, just a suggestion. Works for me either way.
Another questions: should we allow users (who wants to use `BrotliDecoder` directly, not for HTTP) to configure `inputBufferSize`?

Reply:
> Another questions: should we allow users (who wants to use BrotliDecoder directly, not for HTTP) to configure inputBufferSize?

good idea, will do